src/NHibernate/Multi/MultiAnyQueryBase.cs

```diff
@@ -0,0 +1,217 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Linq;
+using NHibernate.Cache;
+using NHibernate.Engine;
+using NHibernate.SqlCommand;
+using NHibernate.Transform;
+using NHibernate.Util;
+
+namespace NHibernate
+{
+	/// <summary>
+	/// Base class for both ICriteria and IQuery queries
+	/// </summary>
+	public abstract partial class MultiAnyQueryBase<TResult> : IMultiAnyQuery, IMultiAnyQuery<TResult>
+	{
+		protected ISessionImplementor Session;
+		private List<object>[] _hydratedObjects;
+		private List<EntityKey[]>[] _subselectResultKeys;
+		private IList[] _loaderResults;
+
+		private List<QueryLoadInfo> _queryInfos;
+		private DbDataReader _reader;
+		private IList<TResult> _finalResults;
+
+		protected class QueryLoadInfo
+		{
+			public Loader.Loader Loader;
+			public QueryParameters Parameters;
+			
+			//Cache related properties:
+			public ISet<string> QuerySpaces;
+			public Action<IList> PutInCacheAction;
+			public CacheableResultTransformer CacheTransformer;
+		}
+
+		protected abstract List<QueryLoadInfo> GetQueryLoadInfo();
+
+		public virtual void Init(ISessionImplementor session)
+		{
+			Session = session;
+
+			_queryInfos = GetQueryLoadInfo();
+
+			var count = _queryInfos.Count;
+			NewArray(count, out _hydratedObjects);
+			NewArray(count, out _subselectResultKeys);
+			NewArray(count, out _loaderResults);
```

Comment:
Why not just writing regular code here?

```c#
_hydratedObjects = new List<object>[count];
_subselectResultKeys = new List<EntityKey[]>[count];
_loaderResults = new IList[count];
```

Adding a general purpose `NewArray` in that class just for this need here looks as a code smell to me. Either remove it or put that as an internal helper on a class like `ArrayHelper`.